Add fallback impl for FileSystem createTemporaryDirectory

serakeri opened the pull request:
#### 00e1e0d6e977c7cc204944834d6af3254a1d5e5b
<pre>
Add fallback impl for FileSystem createTemporaryDirectory
<a href="https://bugs.webkit.org/show_bug.cgi?id=232499">https://bugs.webkit.org/show_bug.cgi?id=232499</a>

Reviewed by Chris Dumez.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::createTemporaryDirectory):
* Source/WTF/wtf/FileSystem.h:

Canonical link: <a href="https://commits.webkit.org/252799@main">https://commits.webkit.org/252799@main</a>
</pre>

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -833,6 +833,33 @@ String pathByAppendingComponents(StringView path, const Vector<StringView>& comp
+    if (!mkdtemp(newTempDir))
+        return String();
+
+    String retval = String::fromUTF8(newTempDir);
```
Should probably use stringFromFileSystemRepresentation().

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -833,6 +833,33 @@ String pathByAppendingComponents(StringView path, const Vector<StringView>& comp
+    std::string newTempDirTemplate = tempDir;
+    newTempDirTemplate += "XXXXXXXX";
+
+    char* newTempDir = new char[newTempDirTemplate.size()];
```
Maybe we could use a Vector<char> to avoid explicit memory management?

serakeri replied:
Thank you for the feedback. I made the changes to the pull request.

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -833,6 +833,30 @@ String pathByAppendingComponents(StringView path, const Vector<StringView>& comp
+    if (!mkdtemp(newTempDir.data()))
+        return String();
+
+    String retval = stringFromFileSystemRepresentation(newTempDir.data());
```
nit: No need for a local variable, just:
```
return stringFromFileSystemRepresentation(newTempDir.data());
```

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -833,6 +833,30 @@ String pathByAppendingComponents(StringView path, const Vector<StringView>& comp
+    if (ec)
+        return String();
+
+    std::string newTempDirTemplate = tempDir;
```
nit: Would this work?
```
std::string newTempDirTemplate = tempDir + "XXXXXXXX";
```

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -833,6 +833,30 @@ String pathByAppendingComponents(StringView path, const Vector<StringView>& comp
+    newTempDirTemplate += "XXXXXXXX";
+
+    Vector<char> newTempDir(newTempDirTemplate.c_str(), newTempDirTemplate.size());
+
```
nit: I don't think we need this new line.

serakeri replied:
I've made the changes and I think it's ready

cdumez reviewed (approved):
LGTM

serakeri committed:
Add fallback impl for FileSystem createTemporaryDirectory
https://bugs.webkit.org/show_bug.cgi?id=232499

Reviewed by Chris Dumez.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::createTemporaryDirectory):
* Source/WTF/wtf/FileSystem.h:

Canonical link: https://commits.webkit.org/252799@main
